src/System.Text.Json/System/Text/Json/JsonWriter.cs

```diff
@@ -420,42 +419,59 @@ private void WriteGuid(Guid value)
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         private void Write(ReadOnlySpan<char> value)
         {
-            var buffer = _output.Buffer;
-            int written;
-            while (!_output.Encoder.TryEncode(value, buffer, out int consumed, out written))
-                buffer = EnsureBuffer();
+            ReadOnlySpan<byte> source = value.AsBytes();
 
-            _output.Advance(written);
-        }
-
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private void WriteJsonValue(ReadOnlySpan<char> values)
-        {
             if (UseFastUtf8)
             {
-                int len = values.Length;
-                var buffer = EnsureBuffer(len);
+                Span<byte> destination = _output.Buffer;
 
-                ref byte utf8Bytes = ref buffer.DangerousGetPinnableReference();
-                ref char chars = ref values.DangerousGetPinnableReference();
-                for (var i = 0; i < len; i++)
-                    Unsafe.Add(ref utf8Bytes, i) = (byte)Unsafe.Add(ref chars, i);
+                while (true)
+                {
+                    var status = Encoders.Utf8.ConvertFromUtf16(source, destination, out int consumed, out int written);
+                    if (status == Buffers.TransformationStatus.Done)
+                    {
+                        _output.Advance(written);
+                        return;
+                    }
+
+                    if (status == Buffers.TransformationStatus.DestinationTooSmall)
+                    {
+                        destination = EnsureBuffer();
+                        continue;
+                    }
+
+                    // This is a failure due to bad input. This shouldn't happen under normal circumstances.
+                    throw new FormatException();
+                }
             }
             else if (UseFastUtf16)
             {
-                int needed = values.Length * sizeof(char);
-                var buffer = EnsureBuffer(needed);
-
-                Span<char> span = buffer.NonPortableCast<byte, char>();
-                values.CopyTo(span);
+                Span<byte> destination = EnsureBuffer(source.Length);
```

Comment:
I think there are some buffers that might not be able to resize to an arbitrary size. i.e. are we sure that EnsureBuffer guarantees the size? I though most of the Ensure calls try the best. Anyway, we need to decide on the contract of "resize" and add tests for it.